src/librbd/ExclusiveLock.cc

```diff
@@ -714,8 +714,11 @@ void ExclusiveLock<I>::handle_shutdown(int r) {
   }
 
   m_image_ctx.aio_work_queue->unblock_writes();
-  m_image_ctx.image_watcher->flush(util::create_context_callback<
-    ExclusiveLock<I>, &ExclusiveLock<I>::complete_shutdown>(this));
+  if (m_image_ctx.image_watcher == nullptr)
```

Comment:
@dillaman I found there are lots of assumption in ExclusiveLock about the image_watcher is not null. seems like the problem is in snap_set(). TBH, I am not clear about the relationship between exclusive_lock and watcher, is that 100% there is at list one watcher if we get an exclusive_lock? Could you help to explain it a bit. thanx a lot.